### PR TITLE
New Buildable: Razorwire

### PR DESCRIPTION
**Pull request by Surrealaser**
:cl: Surrealistik, sprites by Arachnidnexus
add: New buildable obstacle made using 4 metal rods and 1 length of barbed wire.
add: Up to 20 razorwire assemblies per stack; assemblies can fit in the technician's backpack.
add: 3 seconds to set up (-0.5 sec + -0.2 per rank of engi skill), 5 seconds to disassemble with wire cutters (-0.5 sec + -0.5 per rank of engi skill). 8 seconds to repair with welder (-1 sec + -0.5 per rank of engi skill).
add: Blocks movement, but not projectiles, and things that can pass through grilles.
add: Attacking the razorwire in melee results in damage that is reduced by armour.
add: Climbing/pouncing into the razorwire results in significant damage and getting entangled for 5 seconds, as well as an initial slowdown and additional damage on disentangling.
add: Crushers can smash through the razorwire and avoid entanglement but only if they hit it at maximum speed.
add: Boiler acid damages the razorwire.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/razorwire.dm [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+
+/obj/structure/razorwire/attack_tk() // no telehulk sorry
+	return
+
+
+/obj/structure/razorwire/attackby(obj/item/W, mob/user)
+	if(!W)
+		return
+	if(istype(W, /obj/item/grab) && get_dist(src, user) <= 1)
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
	if(istype(W, /obj/item/grab))
```
`attackby` should only be called if adjacent

File: code/game/objects/structures/razorwire.dm [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	if(istype(W, /obj/item/tool/wirecutters))
+		user.visible_message("<span class='notice'>[user] starts disassembling [src].</span>",
+		"<span class='notice'>You start disassembling [src].</span>")
+		var/delay_disassembly = SKILL_TASK_AVERAGE
+		if(user.mind && user.mind.cm_skills && user.mind.cm_skills.engineer) //Higher skill lowers the delay.
+			delay_disassembly -= 5 + user.mind.cm_skills.engineer * 5
+		if(do_after(user,delay_disassembly, TRUE, 5, BUSY_ICON_BUILD))
+			user.visible_message("<span class='notice'>[user] disassembles [src].</span>",
+			"<span class='notice'>You disassemble [src].</span>")
+			playsound(src.loc, 'sound/items/Wirecutter.ogg', 25, 1)
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
			playsound(loc, 'sound/items/Wirecutter.ogg', 25, 1)
```

File: code/game/objects/structures/razorwire.dm [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+	if(istype(W, /obj/item/tool/weldingtool))
+		var/obj/item/tool/weldingtool/WT = W
+		if(WT.remove_fuel(0, user))
+			var/delay = SKILL_TASK_TOUGH
+			if(user.mind && user.mind.cm_skills && user.mind.cm_skills.engineer) //Higher skill lowers the delay.
+				delay -= 10 + user.mind.cm_skills.engineer * 5
+			user.visible_message("<span class='notice'>[user] begins repairing damage to [src].</span>",
+			"<span class='notice'>You begin repairing the damage to [src].</span>")
+			playsound(src.loc, 'sound/items/Welder2.ogg', 25, 1)
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
			playsound(loc, 'sound/items/Welder2.ogg', 25, 1)
```

File: code/game/objects/structures/razorwire.dm [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+			var/old_loc = loc
+			if(do_after(user, delay, TRUE, 5, BUSY_ICON_FRIENDLY) && old_loc == loc)
+				user.visible_message("<span class='notice'>[user] repairs some damage on [src].</span>",
+				"<span class='notice'>You repair [src].</span>")
+				health = min(health + 100, RAZORWIRE_MAX_HEALTH)
+				update_health()
+				playsound(src.loc, 'sound/items/Welder2.ogg', 25, 1)
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
				playsound(loc, 'sound/items/Welder2.ogg', 25, 1)
```

File: code/game/objects/structures/razorwire.dm [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+	if((W.flags_item & ITEM_ABSTRACT) || isrobot(user))
+		return
+
+	var/damage = W.force
+	if(!W.sharp)
+		damage *= 0.25
+	damage = max(damage - soak,0)
+	if(damage)
+		..()
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
		. = ..()
```

File: code/game/objects/structures/razorwire.dm [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+			destroy()
+			return
+		if(2.0)
+			health -= rand(33, 66)
+		if(3.0)
+			health -= rand(10, 33)
+	update_health()
+
+/obj/structure/razorwire/Bumped(atom/A)
+	..()
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
	. = ..()
```

File: code/game/objects/structures/razorwire.dm [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+	if(istype(A, /mob/living/carbon/Xenomorph/Crusher))
+
+		var/mob/living/carbon/Xenomorph/Crusher/C = A
+
+		if(C.charge_speed < C.charge_speed_max/2)
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
		if(C.charge_speed < C.charge_speed_max * 0.5)
```

File: code/game/objects/structures/razorwire.dm [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+			C.visible_message("<span class='danger'>The barbed wire slices into [C]!</span>",
+			"<span class='danger'>The barbed wire slices into you!</span>")
+			C.loc = loc
+			var/armor_block = null
+			var/def_zone = ran_zone()
+			armor_block = C.run_armor_check(def_zone, "melee")
+			C.apply_damage(rand(RAZORWIRE_BASE_DAMAGE * RAZORWIRE_MIN_DAMAGE_MULT_HIGH, RAZORWIRE_BASE_DAMAGE * RAZORWIRE_MAX_DAMAGE_MULT_HIGH), BRUTE, def_zone, armor_block, null, 1) //pouncing into razor wire is especially ouchy
+			C.KnockDown(1)
+			razorwire_tangle(C)
+	..()
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]:
```suggestion
	return ..()
```

File: code/game/objects/structures/razorwire.dm [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+		"<span class='danger'>The barbed wire slices into you!</span>", null, 5)
+		playsound(src, 'sound/effects/barbed_wire_movement.ogg', 25, 1)
+
+
+
+/obj/structure/razorwire/bullet_act(obj/item/projectile/P)
+	bullet_ping(P)
+	health -= round(P.damage * 0.1)
+
+	if(istype(P.ammo, /datum/ammo/xeno/boiler_gas))
```
Comment (Rohesie) [inline code/game/objects/structures/razorwire.dm]: Could open this exception to more xeno attacks, such as different spits and acid spray.

**Comment by Rohesie** [comment]
Will this block the tank?

**Comment by Surrealaser** [comment]
> Will this block the tank?

I don't think so; I didn't notice any problems during the test, nor did anyone mention it.

**Comment by Rohesie** [comment]
If it's dense and an exception is not made for the tank it will block it.

**Pull request merged**